### DISCUSSION TITLE
Call function in file from C

### DISCUSSION THREAD

**Discussion by danimasa** [micropython/micropython#9185] [Using MicroPython]
Hello

How can I call a function from inside a python file from C?

I'm running `mp_compile` but the return value is a function not a dict, and running the function returns a NoneType.

I need to execute a function with a known name from the file.

**Comment by dlech** [comment]
This is done essentially the same way it is done from Python. First you need to import the module that contains the function. This could be done, for example, by calling `mp_builtin___import__` with the appropriate args that are the equivalent of `from module import function`. This will return the function object and you can pass that object to `mp_call_function_*` along with any required args for that function.

**Reply by danimasa** [reply]
I found other function that probably should resolve my problem: `mp_load_name` but after running it I'm getting the following error:

```
Guru Meditation Error: Core  0 panic'ed (LoadProhibited). Exception was unhandled.

Core  0 register dump:
PC      : 0x400e669a  PS      : 0x00060e30  A0      : 0x800df142  A1      : 0x3ffb6c40
0x400e669a: gc_is_locked at /work/modules/micropython/py/gc.c:230

A2      : 0x0000000a  A3      : 0x3ffaf034  A4      : 0x3ffb8b9b  A5      : 0x00000000
A6      : 0x7ff00000  A7      : 0x00000000  A8      : 0x800e669a  A9      : 0x3ffb6c20
A10     : 0x00000000  A11     : 0x3ffb6c7c  A12     : 0x3ffb6c70  A13     : 0x3ffaf034
A14     : 0x00000000  A15     : 0x00000000  SAR     : 0x00000010  EXCCAUSE: 0x0000001c
EXCVADDR: 0x00000008  LBEG    : 0x400014fd  LEND    : 0x4000150d  LCOUNT  : 0xfffffffd


Backtrace: 0x400e6697:0x3ffb6c40 0x400df13f:0x3ffb6c60 0x400d9ce5:0x3ffb6cc0 0x40105271:0x3ffb6ce0 0x40105b71:0x3ffb6dc0 0x40105e19:0x3ffb6e50 0x40105f7c:0x3ffb6e80 0x402202a1:0x3ffb6ea0 0x4011afc5:0x3ffb6ec0 0x400d956e:0x3ffb6ee0 0x400d9cb2:0x3ffb6f50 0x4021efb1:0x3ffb6f70
0x400e6697: gc_is_locked at /work/modules/micropython/py/gc.c:230
```

I don't understand why it's broking in the garbage collector.

**Reply by jimmo** [reply]
@danimasa can you provide more information about what you're doing.

The Python function you want to call? Was that passed in by the user as a callback or something? Or do you have the Python code as a string that you want to execute directly?

Are you invoking this function from something like an interrupt context? Which port are you working in? What calls your C function that you then invoke the Python function from?

**Reply by danimasa** [reply]
> The Python function you want to call? Was that passed in by the user as a callback or something? Or do you have the Python code as a string that you want to execute directly?

I want to call a named function provided by the user in all python files mounted in a FAT filesystem, firstly I want to verify if the function is provided in the files and after that I want to run that function with some parameters and read the response.

> Are you invoking this function from something like an interrupt context? Which port are you working in? What calls your C function that you then invoke the Python function from?

Yes, I'm calling it from a callback that listen to events, In some cases the callback could be called from other task in RTOS, I'm thinking to put a Queue to run all the micropython functions in the same task, but in the example above the callback is running in the same task as micropython. I'm using the ESP32 port.

**Reply by jimmo** [reply]
> I want to call a named function provided by the user in all python files mounted in a FAT filesystem, firstly I want to verify if the function is provided in the files and after that I want to run that function with some parameters and read the response.

OK, so basically this is exactly what `import foo` does (i.e. execute a file `foo.py` from the filesystem, then get things defined in it), so you'll want to do something very similar to `py/builtinimport.c`.

Specifically, this is the line that executes for importing a file from a path. https://github.com/micropython/micropython/blob/master/py/builtinimport.c#L248

Then in do_load_from_lexer, it uses `parse_compile_execute` to run the module with the module's globals dict. In your case you want to probably create your own globals and locals dict.

Then after this runs, any methods defined in that file will now be in the globals dict. You can check if the method was defined by seeing if the required entry (by name) is in the dict. Then execute it with one of the `mp_call_function*` variants.

> In some cases the callback could be called from other task in RTOS

You can only call into the MicroPython VM and runtime from the MicroPython task (or a task that belongs to a MicroPython thread).

**Reply by danimasa** [reply]
I updated the code to work like `py/builtinimport.c` and that part seems to be working, but I'm having trouble with `gc.c` in this line: https://github.com/micropython/micropython/blob/0b26efe73dd3396bdc2b77651a78d9f2edeb9004/py/gc.c#L583

I'm getting the following error: `Guru Meditation Error: Core  0 panic'ed (LoadProhibited). Exception was unhandled.`

What could be the reason for that problem?

**Reply by jimmo** [reply]
This is almost certainly because the code is running somewhere that isn't micropython context... (it's failing trying to access the thread state). Where is this being called from?

**Reply by danimasa** [reply]
It's being called by another task in RTOS, maybe I have to create a specific task for micropython and call it throught a queue. I will try that.

**Reply by dlech** [reply]
Sounds like a job for `mp_sched_schedule()`?

**Reply by jimmo** [reply]
> It's being called by another task in RTOS, maybe I have to create a specific task for micropython and call it throught a queue. I will try that.

Yes like I said above -- "You can only call into the MicroPython VM and runtime from the MicroPython task (or a task that belongs to a MicroPython thread)."

`mp_sched_schedule` is a good option assuming you have some way of passing the required state (remember you cannot allocate on the MicroPython heap from outside a MicroPython task either, although it's fine to use the IDF heap of course). Otherwise you need to have some sort of queue, or a way to synchronise with the MicroPython task (see https://github.com/micropython/micropython/blob/master/extmod/modbluetooth.c#L1191 although I do not necessarily recommend this approach)

**Reply by danimasa** [reply]
In my case I think it's better to run micropython in a different thread, because I don't want to stop the execution of the thread to process the logic written in python. Then I put the micropython in a new task and syncronized it using a queue. But I'm getting a new error when I call the function, I think it's related to my script: `TypeError: unexpected keyword argument '__new__'`.

I don't understand the usage of `mp_sched_schedule`, how do I receive the response from a function I have called? Is it a way to run a function from another thread?

**Reply by danimasa** [reply]
I resolved the error, the problem was I allocated the args in the heap, after putting it in the stack the problem resolved.
